src/Context.php

```diff
@@ -60,11 +64,7 @@ public function enterSibling($pathSegment)
      */
     public function leaveNode()
     {
-        if (count($this->pathSegments) === 0) {
-            throw new \LogicException('Cannot leave node: instance stack is empty');
-        }
```

Comment:
I think this should stay. It's a basic check, and it's almost free with your implementation
(`$this->pathLength === 0`).